src/game/g_stats.c

```diff
@@ -574,6 +574,98 @@ void G_AddKillSkillPointsForDestruction(gentity_t *attacker, meansOfDeath_t mod,
 	CalculateRanks();
 }
 
+#define MAX_PLAYERS_ASSIST_TO_REWARDS 4
+
+/**
+ * @brief G_AddKillAssistPoints
+ * @param[in] target
+ * @param[in] attacker
+ */
+void G_AddKillAssistPoints(gentity_t *target, gentity_t *attacker)
+{
+	unsigned int i;
+	int          bestPlayersAssistReward[MAX_PLAYERS_ASSIST_TO_REWARDS];
+	int          bestTotalDamagePoints[MAX_PLAYERS_ASSIST_TO_REWARDS];
+	qboolean     complaintSent = qfalse;
+
+	if (!target || !target->client)
+	{
+		return;
+	}
+
+	// init arrays to invalid values
+	for (i = 0; i < ARRAY_LEN(bestPlayersAssistReward); ++i)
+	{
+		bestPlayersAssistReward[i] = -1;
+		bestTotalDamagePoints[i]   = -1;
+	}
+
+	// find up to MAX_PLAYERS_ASSIST_TO_REWARDS players to rewards assist points
+	for (i = 0; i < ARRAY_LEN(target->client->damageReceivedFromPlayers); ++i)
+	{
+		// ignore damage received from seld and attacker (killer)
+		if (&g_entities[i] == target || &g_entities[i] == attacker)
+		{
+			continue;
+		}
+
+		// ensure we received damage from the parsed client slot and is from a team different
+		if (target->client->damageReceivedFromPlayers[i] > 0 && g_entities[i].client
+		    && g_entities[i].client->ps.teamNum != TEAM_SPECTATOR)
+		{
+			unsigned int j;
+
+			for (j = 0; j < ARRAY_LEN(bestPlayersAssistReward); ++j)
+			{
+				// free emplacement, use it
+				if (bestPlayersAssistReward[j] < 0)
+				{
+					bestPlayersAssistReward[j] = i;
+					bestTotalDamagePoints[j]   = target->client->damageReceivedFromPlayers[i];
+					break;
+				}
+
+				// check if the previous player deal less damage than the current parsed
+				if (bestTotalDamagePoints[j] < target->client->damageReceivedFromPlayers[i])
+				{
+					// "shift" the table to the right, so we can insert the new best player and erase lower ones
+					if (j < MAX_PLAYERS_ASSIST_TO_REWARDS - 1)
+					{
+						memmove(&bestPlayersAssistReward[j + 1], &bestPlayersAssistReward[j], sizeof(*bestPlayersAssistReward) * (ARRAY_LEN(bestPlayersAssistReward) - j - 1));
+						memmove(&bestTotalDamagePoints[j + 1], &bestTotalDamagePoints[j], sizeof(*bestTotalDamagePoints) * (ARRAY_LEN(bestTotalDamagePoints) - j - 1));
+					}
+
+					bestPlayersAssistReward[j] = i;
+					bestTotalDamagePoints[j]   = target->client->damageReceivedFromPlayers[i];
+					break;
+				}
+			}
+		}
+	}
```

Comment:
I wonder if it would be simpler to have the `damageReceivedFromPlayers` array as a 2D array with `[damage][clientnum]`, if that was the case I believe this whole thing could just be a qsort?